x-pack/plugins/apm/public/components/shared/FilterBar/ServiceEnvFilter.tsx

```diff
@@ -0,0 +1,94 @@
+import { EuiFilterGroup, EuiFilterButton, EuiSelect } from '@elastic/eui';
+import React, { useEffect, useRef, useState } from 'react';
+import { useFetcher } from '../../../hooks/useFetcher';
+import { loadServiceEnvironments } from '../../../services/rest/apm/services';
+import { useLocation } from '../../../hooks/useLocation';
+import { useUrlParams } from '../../../hooks/useUrlParams';
+import { history } from '../../../utils/history';
+import { fromQuery, toQuery } from '../Links/url_helpers';
+
+function without<T>(obj: T, key: string): T {
+  const result: { [objKey: string]: unknown } = Object.assign({}, obj);
+  delete result[key];
+  return result as T;
+}
+
+export const ServiceEnvFilter: React.FC = () => {
+  const location = useLocation();
+  const { urlParams } = useUrlParams();
+  const { start, end, kuery, environment } = urlParams;
+  const { data = [] } = useFetcher(
+    () => loadServiceEnvironments({ start, end, kuery, environment: '' }),
+    [start, end, kuery]
+  );
+  const availableEnvironments = ['Any', ...data, 'Not set'];
+  const [selectedValue, setSelectedValue] = useState(availableEnvironments[0]);
+  const isPostEffect = useRef(false);
+
+  function updateEnvironmentUrl(environment?: string) {
+    if (!environment || environment === availableEnvironments[0]) {
```

Comment:
Can you extract the `if` to make it more readable?
```js
const allEnvironments = !environment || environment === allEnvironmentsKey
if (!environment || environment === availableEnvironments[0]) {
```

Btw. how is it possible to select no environment (`!environment`)?

Reply:
>Btw. how is it possible to select no environment (!environment)?

I was guarding for the case when environment is not in the URL, along with the other part of the conditional, it asserts that the URL is `'Any'` or just not defined. But it seems like it's unnecessary since it would essentially be a noop when it's no defined. Regardless, the bulk of this component is to be replaced.

Reply:
> I was guarding for the case when environment is not in the URL

I don't quite get this. Isn't `updateEnvironmentUrl` called `onChange` ? In that case something was selected, right?